content/blog/sharp-edges-in-rust-std/index.md

```diff
@@ -0,0 +1,207 @@
+For new code, I recommend using [`thread::scope`](https://doc.rust-lang.org/std/thread/fn.scope.html) instead, which is a much better API in every conceivable way.
+The documentation addresses the above issue directly:
+
+> Unlike non-scoped threads, scoped threads can borrow non-`'static` data, as the scope guarantees **all threads will be joined at the end of the scope**.
+> All threads spawned within the scope that **haven't been manually joined will be automatically joined** before this function returns.
+
+Alternatively, you can use a thread pool library or [rayon](https://github.com/rayon-rs/rayon), in case you have an iterator you want to parallelize to many threads.
+
+## `std::collections::LinkedList`
+
+Implementing a linked list in Rust is [not easy](https://rust-unofficial.github.io/too-many-lists/).
+That's because Rust's ownership model is detrimental to self-referential data structures. 
+Some people might not know that the standard library ships with an implementation of a linked list at [`std::collections::LinkedList`](https://doc.rust-lang.org/std/collections/struct.LinkedList.html).
+
+In all those years, I never felt the urge to use it. 
+It might even be the least-used collection type in the standard library overall.
+
+One reason might be that for all ordinary use cases, a `Vec` is sufficient and straightforward to use.
+Vectors also have better cache locality and performance characteristics: all items are stored in contiguous memory, which is much better for fast memory access.
+On the other side, elements in a linked list can be scattered all over the heap.
+If you want to learn more, you can read this paper titled ["RIP Linked List - Empirical Study to Discourage You from Using Linked Lists Any Further"](https://arxiv.org/pdf/2306.06942), which contains some benchmarks.
+
+You might be wondering why linked lists get used at all.
+They have their place as a very specialized data structure that is only really helpful in some very low-level environments like a kernel. The Linux kernel uses a lot of linked lists, for example. The reason is that they are very efficient for inserting and deleting elements at arbitrary positions, which is great when you write things like a kernel scheduler.
+
+As for normal, everyday code, just use a `Vec`.
+Even [the documentation of `LinkedList`](https://doc.rust-lang.org/nightly/std/collections/struct.LinkedList.html) itself agrees:
+
+> NOTE: It is almost always better to use `Vec` or `VecDeque` because array-based
+containers are generally faster, more memory efficient, and make better use of
+CPU cache.
+
+I believe the LinkedList should not have been included in the standard library in the first place.
+There is a [longer discussion in the Rust forum](https://internals.rust-lang.org/t/whats-the-status-of-std-linkedlist-maybe-deprecate-in-rust-2018/8068).
+
+## `std::collections::BTreeMap`
+
+That one's a bit controversial.
+
+Rust has two hash map implementations in the standard library:
```

Comment:
BTree doesn't use hashing, so I think it's more accurate to say "has two map implementations".

Reply:
fixed